### ISSUE TITLE
[iOS] OnResume is being called twice when the application is foregrounded

### ISSUE THREAD

**Issue by adrianknight89** [xamarin/Xamarin.Forms#4608]
### Steps to Reproduce

1. Run the attached repro
2. Background and foreground the app and observe that the counter is output to the console
3. Background and foreground the app several times until you see OnResume fire twice in succession and the counter get updated twice

### Expected Behavior
OnResume should fire only once (Please also test OnSleep. I haven't tested it, but we need to make sure that one is safe also.)

### Reproduction Link

[App1.zip](https://github.com/xamarin/Xamarin.Forms/files/2636523/App1.zip)

**Comment by kingces95** [comment]
iOS works for me. Tried a dozen times. Only called once. The 82322 comment says, "I'm using 2.3.2.127". The 32272 is reporting for Android. 59980 reports the Android issue but, like Paul, I'm unable to reproduce it. 12/12 restarts and counters on both iOS XS 12.1/Android.

I believe you're seeing this, but we'll need a reproduction to fix it... How many restarts do you have to preform before it manifests?

**Comment by adrianknight89** [comment]
@kingces95 Sorry, I forgot to mention that the issue happens when you click the app icon on the home screen when foregrounding the app, and you only need to do this a few times to reproduce the issue. I updated the sample code so that the counter is displayed on the screen. I provided the other URLs since people had a similar issue in the past (not because they are necessarily the same issue as the one I reported).

[App1 (updated).zip](https://github.com/xamarin/Xamarin.Forms/files/2658556/App1.updated.zip)

**Comment by adrianknight89** [comment]
Yeah, it should be that way, but I reproduced it on an actual device - not a simulator.

**Comment by kingces95** [comment]
@samhouts Mr Knight says I gotta go buy myself a spanking new iPhone to reproduce this issue... ;)

**Comment by samhouts** [comment]
Well, I never saw the counter increment at all! It just stayed at 0. Tested on an iPhone XS Max device.

**Comment by adrianknight89** [comment]
It doesn't make sense. The counter should increment when `OnResume` is called. Are you backgrounding and foregrounding the app (on a physical device)? Are you testing the second zip I provided? Please also make sure to tap the app icon every time the app needs to be shown.

**Comment by samhouts** [comment]
Yep! I am backgrounding the app and resuming by tapping on the icon on a physical device. It just stays 0. Sometimes the splash background is all that I see. Very odd.

**Comment by adrianknight89** [comment]
Weird. You might want to uninstall and install the app again?

**Comment by adrianknight89** [comment]
Here's a test run I made on XS:

https://1drv.ms/v/s!AjlbPgOcTyP2cJq5iWTtFqgMBMk

Notice that after 6, `OnResume` fired once, but every other time it fired twice.

Also, timing seems to matter somewhat. If you wait long enough between transitions, `OnResume` fires once, but if you switch back and forth quickly. it fires twice.

**Comment by adrianknight89** [comment]
@samhouts I did some more testing and found out that the reason why `OnResume` is getting called twice has to do with `DidEnterBackground`. If you send the app to the background, it takes 1-2 seconds before `DidEnterBackground` is called (tested on XS). If you bring it back to the foreground before `DidEnterBackground` is called, then `OnResume` fires twice.

Wait 5 seconds before bringing the app to the foreground:

```
[0:] OnSleep (XF)
[0:] OnResignActivation (Native)
[0:] DidEnterBackground: Background (Native)
[0:] WillEnterForeground (Native)
[0:] OnResume (XF)
[0:] OnActivated (Native)
```
Rapid state switching does not call `DidEnterBackground`, hence twice the number of events:

```
[0:] OnSleep
[0:] OnResignActivation
[0:] OnResume
[0:] OnActivated
[0:] 
[0:] 
[0:] OnSleep
[0:] OnResignActivation
[0:] OnResume
[0:] OnActivated
```
This is also the reason why `WillEnterForeground ` is not called.
